In [1]:
from google.colab import files
uploaded = files.upload()  # اختاري الـ zip من جهازك

Saving My First Project.v1-testing.yolov8.zip to My First Project.v1-testing.yolov8.zip


In [2]:
import zipfile, os

zip_name = list(uploaded.keys())[0]
with zipfile.ZipFile(zip_name, 'r') as zip_ref:
    zip_ref.extractall("/content/dataset")


os.listdir("/content/dataset")

['README.roboflow.txt', 'README.dataset.txt', 'train', 'test', 'data.yaml']

In [3]:
!find /content/dataset -name "data.yaml"

/content/dataset/data.yaml


In [4]:
!cat /content/dataset/data.yaml

train: ../train/images
val: ../valid/images
test: ../test/images

nc: 9
names: ['battery', 'electronic-waste', 'keyboard', 'laptop', 'mobile phone', 'monitor', 'mouse', 'pcb', 'printer']

roboflow:
  workspace: dataset-ypcll
  project: my-first-project-thxpx
  version: 1
  license: CC BY 4.0
  url: https://universe.roboflow.com/dataset-ypcll/my-first-project-thxpx/dataset/1

In [5]:
data_yaml_content = """
train: /content/dataset/train/images
val: /content/dataset/valid/images
test: /content/dataset/test/images

nc: 9
names: ['battery', 'electronic-waste', 'keyboard', 'laptop', 'mobile phone', 'monitor', 'mouse', 'pcb', 'printer']
"""

with open("/content/dataset/data.yaml", "w") as f:
    f.write(data_yaml_content)

In [7]:
!find /content/dataset -maxdepth 3

Streaming output truncated to the last 5000 lines.
/content/dataset/train/labels/e-waste-1717-_jpg.rf.57a81007e4ba91b43499983dd99e6322.txt
/content/dataset/train/labels/e-waste-2301-_jpg.rf.c42d6e9be95b2ac21da513c75ef0d3f9.txt
/content/dataset/train/labels/e-waste-3283-_jpg.rf.cce1cc909d921355397a6d36f43bd062.txt
/content/dataset/train/labels/e-waste-2577-_jpg.rf.193f2234b40aed04b54a2d25dbaf5f0b.txt
/content/dataset/train/labels/e-waste-3792-_jpg.rf.3855525bd4524fb2c249de833f9820e5.txt
/content/dataset/train/labels/e-waste-5516-_jpg.rf.7de9a72da0bf1d044c8218453cd1dd4e.txt
/content/dataset/train/labels/e-waste-7413-_jpg.rf.89737d806565f7e5ab6a5f55ac0ba6de.txt
/content/dataset/train/labels/e-waste-4071-_jpg.rf.4a781acf2c12829ae8c9cb80e2b99c7f.txt
/content/dataset/train/labels/e-waste-379-_jpg.rf.fd2fb382bc739b027be3d80055f85917.txt
/content/dataset/train/labels/e-waste-3951-_jpg.rf.60395fe58ef620c6fa8b703c1e663a3c.txt
/content/dataset/train/labels/e-waste-4976-_jpg.rf.5460655a5558d99282b

In [8]:
data_yaml_content = """
train: /content/dataset/train/images
val: /content/dataset/test/images
test: /content/dataset/test/images

nc: 9
names: ['battery', 'electronic-waste', 'keyboard', 'laptop', 'mobile phone', 'monitor', 'mouse', 'pcb', 'printer']
"""

with open("/content/dataset/data.yaml", "w") as f:
    f.write(data_yaml_content)

print(open("/content/dataset/data.yaml").read())


train: /content/dataset/train/images
val: /content/dataset/test/images
test: /content/dataset/test/images

nc: 9
names: ['battery', 'electronic-waste', 'keyboard', 'laptop', 'mobile phone', 'monitor', 'mouse', 'pcb', 'printer']



In [10]:
!pip install ultralytics
from ultralytics import YOLO

model = YOLO("yolov8n.pt")

results = model.train(
    data="/content/dataset/data.yaml",
    epochs=25,
    imgsz=640,
    batch=16,
    name="ewaste_yolo",
)

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 464.7 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.7/88.7 kB 10.9 MB/s eta 0:00:00
Creating new Ultralytics Settings v0.0.8 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/usage/settings.
Ultralytics 8.4.159 🚀 Python-3.13.15 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, channels_last=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, cls_remap=True, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/dataset/data.yaml, degrees=0.0, deterministic=

In [11]:
import os

train_labels = "/content/dataset/train/labels"
class_names = ['battery', 'electronic-waste', 'keyboard', 'laptop', 'mobile phone', 'monitor', 'mouse', 'pcb', 'printer']

counts = {name: 0 for name in class_names}

for file in os.listdir(train_labels):
    with open(os.path.join(train_labels, file)) as f:
        for line in f:
            class_id = int(line.split()[0])
            counts[class_names[class_id]] += 1

for name, count in counts.items():
    print(f"{name}: {count}")

battery: 1244
electronic-waste: 852
keyboard: 515
laptop: 734
mobile phone: 1374
monitor: 584
mouse: 623
pcb: 479
printer: 185


In [12]:
test_labels = "/content/dataset/test/labels"
class_names = ['battery', 'electronic-waste', 'keyboard', 'laptop', 'mobile phone', 'monitor', 'mouse', 'pcb', 'printer']

counts = {name: 0 for name in class_names}

for file in os.listdir(test_labels):
    with open(os.path.join(test_labels, file)) as f:
        for line in f:
            class_id = int(line.split()[0])
            counts[class_names[class_id]] += 1

for name, count in counts.items():
    print(f"{name}: {count}")

battery: 47
electronic-waste: 100
keyboard: 54
laptop: 67
mobile phone: 43
monitor: 58
mouse: 76
pcb: 0
printer: 0


In [15]:
import os
import shutil
import random
from collections import defaultdict

random.seed(42)

base = "/content/dataset"
all_images_dir = "/content/dataset_combined/images"
all_labels_dir = "/content/dataset_combined/labels"
os.makedirs(all_images_dir, exist_ok=True)
os.makedirs(all_labels_dir, exist_ok=True)

for split in ["train", "test"]:
    img_dir = os.path.join(base, split, "images")
    lbl_dir = os.path.join(base, split, "labels")
    for f in os.listdir(img_dir):
        shutil.copy(os.path.join(img_dir, f), all_images_dir)
    for f in os.listdir(lbl_dir):
        shutil.copy(os.path.join(lbl_dir, f), all_labels_dir)

print("photos :", len(os.listdir(all_images_dir)))

photos : 7026


In [19]:
import os
import shutil
import random
from collections import defaultdict

random.seed(42)

images_dir = "/content/dataset_combined/images"
labels_dir = "/content/dataset_combined/labels"

class_names = ['battery', 'electronic-waste', 'keyboard', 'laptop', 'mobile phone',
               'monitor', 'mouse', 'pcb', 'printer']

label_files = [f for f in os.listdir(labels_dir) if f.endswith(".txt")]

class_to_files = defaultdict(list)
for lf in label_files:
    with open(os.path.join(labels_dir, lf)) as f:
        classes_in_file = set(int(line.split()[0]) for line in f if line.strip())
    if not classes_in_file:
        continue
    rarest_class = min(classes_in_file, key=lambda c: len(class_to_files[c]) if c in class_to_files else 0)
    class_to_files[rarest_class].append(lf)

train_files = set()
test_files = set()

for cls, files in class_to_files.items():
    random.shuffle(files)
    split_idx = int(len(files) * 0.8)
    train_files.update(files[:split_idx])
    test_files.update(files[split_idx:])

print("Number of train files:", len(train_files))
print("Number of test files:", len(test_files))


Number of train files: 5617
Number of test files: 1408


In [20]:
new_base = "/content/dataset_v2"
new_train_img = os.path.join(new_base, "train/images")
new_train_lbl = os.path.join(new_base, "train/labels")
new_test_img = os.path.join(new_base, "test/images")
new_test_lbl = os.path.join(new_base, "test/labels")

for d in [new_train_img, new_train_lbl, new_test_img, new_test_lbl]:
    os.makedirs(d, exist_ok=True)

def get_image_filename(label_filename):
    base_name = os.path.splitext(label_filename)[0]
    for ext in [".jpg", ".jpeg", ".png"]:
        candidate = base_name + ext
        if os.path.exists(os.path.join(images_dir, candidate)):
            return candidate
    return None

for lf in train_files:
    img_file = get_image_filename(lf)
    if img_file:
        shutil.copy(os.path.join(images_dir, img_file), new_train_img)
        shutil.copy(os.path.join(labels_dir, lf), new_train_lbl)

for lf in test_files:
    img_file = get_image_filename(lf)
    if img_file:
        shutil.copy(os.path.join(images_dir, img_file), new_test_img)
        shutil.copy(os.path.join(labels_dir, lf), new_test_lbl)

print("Train images copied:", len(os.listdir(new_train_img)))
print("Test images copied:", len(os.listdir(new_test_img)))

Train images copied: 5617
Test images copied: 1408


In [21]:
new_test_labels = "/content/dataset_v2/test/labels"

counts = {name: 0 for name in class_names}

for file in os.listdir(new_test_labels):
    with open(os.path.join(new_test_labels, file)) as f:
        for line in f:
            class_id = int(line.split()[0])
            counts[class_names[class_id]] += 1

for name, count in counts.items():
    print(f"{name}: {count}")

battery: 258
electronic-waste: 191
keyboard: 114
laptop: 162
mobile phone: 285
monitor: 129
mouse: 140
pcb: 96
printer: 37


In [22]:
yaml_content = """train: /content/dataset_v2/train/images
val: /content/dataset_v2/test/images
test: /content/dataset_v2/test/images

nc: 9
names: ['battery', 'electronic-waste', 'keyboard', 'laptop', 'mobile phone', 'monitor', 'mouse', 'pcb', 'printer']
"""

with open("/content/dataset_v2/data.yaml", "w") as f:
    f.write(yaml_content)

print("data.yaml created at /content/dataset_v2/data.yaml")

data.yaml created at /content/dataset_v2/data.yaml


In [23]:
from ultralytics import YOLO

model = YOLO("yolov8n.pt")

results = model.train(
    data="/content/dataset_v2/data.yaml",
    epochs=25,
    imgsz=640,
    batch=16,
    name="ewaste_yolo_v2",
)

Ultralytics 8.4.159 🚀 Python-3.13.15 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, channels_last=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, cls_remap=True, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/dataset_v2/data.yaml, degrees=0.0, deterministic=True, device=, dfl=1.5, dgrad=0.5, dis=6.0, distill_model=None, dlam=1.0, dlog=1.0, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=25, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=ewaste_yolo_v2, nbs=64, nms=None, opset=None, o

## 2.6 Evaluation — YOLO Vision Component

### Initial Split (v1)
The first train/test split was a simple random split, which resulted in **2 out of 9 classes (`pcb`, `printer`) having zero instances in the test set** — despite being present in training (479 and 185 instances respectively). This made it impossible to fairly evaluate the model on these classes.

Initial results (random split):
- mAP50: 0.399
- mAP50-95: 0.282
- Several classes (`battery`, `mouse`, `electronic-waste`) scored very low (mAP50 < 0.12), largely due to insufficient/unbalanced test representation.

### Fix: Stratified Split (v2)
To address this, a stratified split was implemented: each image was assigned to a class bucket based on its rarest annotated class, then split 80/20 within each bucket. This ensured all 9 classes were represented in both train and test sets.

### Final Results (v2, YOLOv8n, 25 epochs)

| Class | Images | Instances | Precision | Recall | mAP50 | mAP50-95 |
|---|---|---|---|---|---|---|
| all | 1408 | 1412 | 0.910 | 0.869 | 0.920 | 0.663 |
| battery | 258 | 258 | 0.908 | 0.888 | 0.916 | 0.721 |
| electronic-waste | 191 | 191 | 0.853 | 0.788 | 0.893 | 0.702 |
| keyboard | 114 | 114 | 0.964 | 0.947 | 0.964 | 0.637 |
| laptop | 159 | 162 | 0.954 | 0.887 | 0.952 | 0.584 |
| mobile phone | 285 | 285 | 0.979 | 0.947 | 0.984 | 0.735 |
| monitor | 129 | 129 | 0.912 | 0.891 | 0.951 | 0.686 |
| mouse | 140 | 140 | 0.937 | 0.952 | 0.979 | 0.690 |
| pcb | 96 | 96 | 0.779 | 0.740 | 0.748 | 0.551 |
| printer | 37 | 37 | 0.906 | 0.778 | 0.893 | 0.657 |

### Failure Case Analysis
The stratified split alone improved overall mAP50 from **0.399 to 0.920** (+0.521) and mAP50-95 from **0.282 to 0.663**, confirming the original bottleneck was data split quality, not model capacity.

The weakest remaining class is `pcb` (mAP50 0.748), likely due to:
- Smallest instance count among all classes (96, vs. 258+ for top classes)
- PCBs vary heavily in visual appearance (size, color, component layout) compared to more visually consistent classes like `monitor` or `mobile phone`

`laptop` also lags slightly on mAP50-95 (0.584), suggesting looser bounding-box precision — possibly due to laptops often being partially occluded or photographed at varying angles.

**Mitigation applied:** stratified splitting to guarantee class representation across train/test.
**Future improvement (not applied due to time constraints):** additional augmentation or oversampling for `pcb` and `printer`, or more training epochs.

In [30]:
!cp runs/detect/ewaste_yolo_v2/weights/best.pt /content/best_yolo.pt